src/sentry/event_manager.py

```diff
@@ -690,6 +697,57 @@ def _pull_out_data(jobs: Sequence[Job], projects: ProjectsMapping) -> None:
         job["groups"] = []
 
 
+def _set_project_platform_if_needed(project: Project, event: Event) -> None:
+    # Only infer the platform if it's useful - if the event platform is "other" or null, there's
+    # no useful information for us to set the project platform
+    if not event.platform or event.platform == "other":
+        return
+
+    # Use a lock to prevent race conditions when multiple events are processed
+    # concurrently for a project with no initial platform
+    lock_key = f"project-platform-lock:{project.id}"
+
+    try:
+        with locks.get(lock_key, duration=60, name="project-platform-lock").acquire():
+            project.refresh_from_db(fields=["platform"])
+
+            if not project.platform:
+                with transaction.atomic(router.db_for_write(Project)):
+                    project.update(platform=event.platform)
+                    project.update_option("sentry:project_platform_inferred", event.platform)
+
+                    create_system_audit_entry(
+                        organization=project.organization,
+                        target_object=project.id,
+                        event=audit_log.get_event_id("PROJECT_EDIT"),
+                        data={**project.get_audit_log_data(), "platform": event.platform},
+                    )
+                return
+
+            if project.platform != event.platform:
+                inferred_platform = project.get_option("sentry:project_platform_inferred")
+
+                # If current platform matches what we inferred, we can safely continue inferring it
+                # since it hasn't been manually changed
+                if inferred_platform and project.platform == inferred_platform:
+                    with transaction.atomic(router.db_for_write(Project)):
+                        project.update(platform="other")
+                        project.update_option("sentry:project_platform_inferred", "other")
+                        create_system_audit_entry(
+                            organization=project.organization,
+                            target_object=project.id,
+                            event=audit_log.get_event_id("PROJECT_EDIT"),
+                            data={**project.get_audit_log_data(), "platform": "other"},
+                        )
+
+                # Otherwise, we need to mark that we should stop inferring platform (unless it becomes null again)
+                else:
+                    project.update_option("sentry:project_platform_inferred", None)
+
+    except Exception:
+        return
```

Comment:
We don't expect errors, so if we do get one, we'd like to know what it is. Let's capture it in Sentry (which we get for free if we use `logger.exception`).